bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/handler/AbstractSunSpecHandler.java

```diff
@@ -0,0 +1,495 @@
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.library.types.QuantityType;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.State;
+import org.eclipse.smarthome.core.types.UnDefType;
+import org.openhab.binding.modbus.handler.EndpointNotInitializedException;
+import org.openhab.binding.modbus.handler.ModbusEndpointThingHandler;
+import org.openhab.binding.modbus.sunspec.internal.SunSpecConfiguration;
+import org.openhab.binding.modbus.sunspec.internal.dto.ModelBlock;
+import org.openhab.io.transport.modbus.BasicModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.BasicPollTaskImpl;
+import org.openhab.io.transport.modbus.BitArray;
+import org.openhab.io.transport.modbus.ModbusManager;
+import org.openhab.io.transport.modbus.ModbusReadCallback;
+import org.openhab.io.transport.modbus.ModbusReadFunctionCode;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.openhab.io.transport.modbus.PollTask;
+import org.openhab.io.transport.modbus.endpoint.ModbusSlaveEndpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link AbstractSunSpecHandler} is the base class for any sunspec handlers
+ * Common things are handled here:
+ *
+ * - loads the configuration either from the configuration file or
+ * from the properties that have been set by the auto discovery
+ * - sets up a regular poller to the device
+ * - handles incoming messages from the device:
+ * - common properties are parsed and published
+ * - other values are submitted to child implementations
+ * - handles disposal of the device by removing any handlers
+ * - implements some tool methods
+ *
+ * @author Nagy Attila Gabor - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractSunSpecHandler extends BaseThingHandler {
+
+    /**
+     * Logger instance
+     */
+    private final Logger logger = LoggerFactory.getLogger(AbstractSunSpecHandler.class);
+
+    /**
+     * Configuration instance
+     */
+    protected Optional<SunSpecConfiguration> config = Optional.empty();
+
+    /**
+     * This is the task used to poll the device
+     */
+    private volatile Optional<PollTask> pollTask = Optional.empty();
+
+    /**
+     * This is the slave endpoint we're connecting to
+     */
+    protected volatile Optional<ModbusSlaveEndpoint> endpoint = Optional.empty();
```

Comment:
I think using `@Nullable` instead would simplify your code very much.

Reply:
@J-N-K I disagree here, and it's not just about coding taste.

My problem is with the Eclipse null checker, wich when using @Nullable types gives a dozens of false positives.

More annoyingly it doesn't seem to be coherent: some solutions work at one place, but not at the other. And once I fix everything in the Eclipse IDE and try to compile with maven I get errors again for lines that passed in the IDE or vica versa.

As you requested I've replaced the pollTask with a @Nullable, and now I had to write such ugly hacks:
```java
PollTask task = (@NonNull PollTask) pollTask;
managerRef.unregisterRegularPoll(task);
```
because writing this didn't satisfy the null checker:
```java
if (pollTask != null) {
    managerRef.unregisterRegularPoll(task);
}
```

So after pulling most of my remaining hair I've took a big breath and really replaced everything with @Nonnull Optionals. And while first this seems to be an overkill, during the journey I begun to like it: it's much more talkative, it's easier to recognize where to do the necessary checks, and also it gives a nice fluent api to substitute missing values with defaults.

Reply:
I wouldn't say that the null checker is wrong (except when it comes to maps) just a bit overzealous. Almost all of the null check warnings can be solved by caching fields to a local variable and then operating on the local variable. The reason that the null checker still complains even after you do a null check is that in a concurrent environment another thread could change your field values as you do your null check. Now normally this isn't the case, but the null checker doesn't know that and it is erring on the side of caution.

So the way that you would solve your example would be to do this:
```
PollTask task = pollTask;
if(task != null) {
    managerRef.unregisterRegularPoll(task);
}
```

Reply:
Yes, I understand this, but then I get the warning from the java compiler, that the given variable can not be null there. Or the IDE keeps removing my lines, because it thinks it's pointless... and this is where I'm starting to pull my hair out.
At the end of day I thinks it's a matter of taste if you like the Optional better or the local variable solution.

Anyway, if this the only thing that blocks my PR being merged, please let me know, and I'll rewrite it.

Thanks for the review.